<h1 class="main-heading">Macro std::try</h1>
<pre class="rust item-decl" data-language="rust">macro_rules! try {
    ($expr:expr $(,)?) =&gt; { ... };
}</pre>
<span class="item-info"><div class="stab deprecated">
<span class="emoji">ðŸ‘Ž</span><span>Deprecated since 1.39.0: use the <code>?</code> operator instead</span>
</div></span><p>Unwraps a result or propagates its error.</p> <p>The <a href="../book/ch09-02-recoverable-errors-with-result.html#a-shortcut-for-propagating-errors-the--operator"><code>?</code> operator</a> was added to replace <code>try!</code> and should be used instead. Furthermore, <code>try</code> is a reserved word in Rust 2018, so if you must use it, you will need to use the <a href="https://doc.rust-lang.org/rust-by-example/compatibility/raw_identifiers.html">raw-identifier syntax</a>: <code>r#try</code>.</p> <p><code>try!</code> matches the given <a href="result/enum.result.html" title="enum std::result::Result"><code>Result</code></a>. In case of the <code>Ok</code> variant, the expression has the value of the wrapped value.</p> <p>In case of the <code>Err</code> variant, it retrieves the inner error. <code>try!</code> then performs conversion using <code>From</code>. This provides automatic conversion between specialized errors and more general ones. The resulting error is then immediately returned.</p> <p>Because of the early return, <code>try!</code> can only be used in functions that return <a href="result/enum.result.html" title="enum std::result::Result"><code>Result</code></a>.</p> <h2 id="examples">Examples</h2> <div class="example-wrap"><pre class="rust rust-example-rendered" data-language="rust">use std::io;
use std::fs::File;
use std::io::prelude::*;

enum MyError {
    FileWriteError
}

impl From&lt;io::Error&gt; for MyError {
    fn from(e: io::Error) -&gt; MyError {
        MyError::FileWriteError
    }
}

// The preferred method of quick returning Errors
fn write_to_file_question() -&gt; Result&lt;(), MyError&gt; {
    let mut file = File::create("my_best_friends.txt")?;
    file.write_all(b"This is a list of my best friends.")?;
    Ok(())
}

// The previous method of quick returning Errors
fn write_to_file_using_try() -&gt; Result&lt;(), MyError&gt; {
    let mut file = r#try!(File::create("my_best_friends.txt"));
    r#try!(file.write_all(b"This is a list of my best friends."));
    Ok(())
}

// This is equivalent to:
fn write_to_file_using_match() -&gt; Result&lt;(), MyError&gt; {
    let mut file = r#try!(File::create("my_best_friends.txt"));
    match file.write_all(b"This is a list of my best friends.") {
        Ok(v) =&gt; v,
        Err(e) =&gt; return Err(From::from(e)),
    }
    Ok(())
}</pre></div><div class="_attribution">
  <p class="_attribution-p">
    &copy; 2010 The Rust Project Developers<br>Licensed under the Apache License, Version 2.0 or the MIT license, at your option.<br>
    <a href="https://doc.rust-lang.org/std/macro.try.html" class="_attribution-link">https://doc.rust-lang.org/std/macro.try.html</a>
  </p>
</div>
